### ISSUE TITLE
Numerical instability somewhere in SCF_GUESS RESTART

### ISSUE THREAD

**Issue by oschuett** [cp2k/cp2k#114]
I believe this race affects at least these tests:

-  `QS/regtest-gpw-5/si7c_broy_gapw_a04_restart.inp`
-  `QS/regtest-gapw-1/c8_broy_gapw_all_xashh.inp`
- `QS/regtest-gpw-7/si8_broy_kry_r.inp`

But probably there are more. After we fix this we should revisit the thresholds of all tests that use `SCF_GUESS RESTART`

The good news is that it's easy to reproduce as it shows up in about 10% of the runs:
```
for i in {1..10} ; do 
  ../../../exe/local/cp2k.ssmp si7c_broy_gapw_a04_restart.inp | grep -e "Total energy:"
done
```

**Comment by oschuett** [comment]
Unfortunately, the problem persists even when compiled with `-finit-derived -finit-local-zero`.

**Comment by mkrack** [comment]
Finally, I could reproduce a wrong result for the test input

`QS/regtest-gpw-5/si7c_broy_gapw_a04_restart.inp`

using the CP2K toolchain when I start with the gcc-8.2.0 compiler and build the toolchain with

`./install_cp2k_toolchain.sh --with-gcc --with-mpich --install-all`

Already the arch file local.sdbg gives reproducibly a wrong result in the order of 1.0E-6. At a first glance that seems to originate from differences in the information read from the restart file, but that turned out to be wrong, though the MO eigenvectors may differ depending on the built. But this is due to MO degeneracy which causes differences in the MO orthonormalization.
Tracing the issue led to the routine make_basis_lowdin which builds (C^T *S*C)^(-1/2) for the Lowdin orthonormalisation of the MOs read from the restart file in the case of smearing. Though the MO coefficient C from the restart file were already orthonormal for the given overlap matrix S, the routine does not return a unit matrix. This is caused by a not fully orthonormal set of eigenvectors returned by the LAPACK solver routine SYEVD. For the specific case above, only one pair of eigenvectors is not orthonormal (0.0044 instead of 0.0), but that is sufficient to spoil the transformation of  S. Thus, eventually, the originally orthonormal MO eigenvectors become slightly non-orthonormal after that "re-orthonomalisation step".
I have introduced a CHECK_DIAG flag in my cp2k fork that checks the orthonormality of the eigenvectors calculated by the routine `choose_eigv_solver`. A full regression tester run with local.sdbg resulted in failures for the following test inputs:
- QS/regtest-gpw-5/si7c_broy_gapw_a04_restart.inp
- Fist/regtest-opt/cell_opt_direct_bfgs.inp
- Fist/regtest-opt/cs_geo_opt_bfgs.inp
- Fist/regtest-opt/cs_cell_opt_bfgs_geo_opt_bfgs.inp
- Fist/regtest-12/argon_opt_cell.inp
- Fist/regtest-12/argon_opt_cell_ks.inp
- Fist/regtest-7-2/UO2-2x2x2-genpot_units.inp
- SE/regtest-2-1/md.inp
Thus this is obviously not an OpenMP or parallelisation issue. I did not observe such the problem with other regression testers, e.g. using gcc-6.4.0, so far. Therefore, it is still possible that the problem only appears when LAPACK/SCALAPACK is compiled with gcc-8.2.0. I will try to merge my changes to the trunk hoping to get more information.

**Comment by oschuett** [comment]
Unfortunately, the problem seems to [persist](https://dashboard.cp2k.org/archive/current-sopt/commit_1bc924c34d481b1d2b3916ceeb1dd56afad2393b.txt) :-/

**Comment by alazzaro** [comment]
This numerical instabilities are similar to what we see in #1 ...

**Comment by oschuett** [comment]
I believe this got resolved by dropping `-ffast-math` via #195.
